cpp/src/io/orc/reader_impl.hpp

```diff
@@ -91,143 +63,27 @@ class reader::impl {
-   * @param level Current nesting level.
-   */
-  column_buffer&& assemble_buffer(size_type const orc_col_id,
-                                  std::vector<std::vector<column_buffer>>& col_buffers,
-                                  size_t const level,
-                                  rmm::cuda_stream_view stream);
-
-  /**
-   * @brief Create columns and respective schema information from the buffer.
-   *
-   * @param col_buffers Column buffers for columns and children.
-   * @param out_columns Vector of columns formed from column buffers.
-   * @param schema_info Vector of schema information formed from column buffers.
-   * @param stream CUDA stream used for device memory operations and kernel launches
-   */
-  void create_columns(std::vector<std::vector<column_buffer>>&& col_buffers,
-                      std::vector<std::unique_ptr<column>>& out_columns,
-                      std::vector<column_name_info>& schema_info,
-                      rmm::cuda_stream_view stream);
-
-  /**
-   * @brief Create empty columns and respective schema information from the buffer.
-   *
-   * @param col_buffers Column buffers for columns and children.
-   * @param schema_info Vector of schema information formed from column buffers.
-   * @param stream CUDA stream used for device memory operations and kernel launches
-   *
-   * @return An empty column equivalent to orc column type.
-   */
-  std::unique_ptr<column> create_empty_column(size_type const orc_col_id,
-                                              column_name_info& schema_info,
-                                              rmm::cuda_stream_view stream);
-
-  /**
-   * @brief Setup table for converting timestamp columns from local to UTC time
-   *
-   * @return Timezone table with timestamp offsets
-   */
-  std::unique_ptr<table> compute_timezone_table(
-    std::vector<cudf::io::orc::metadata::stripe_source_mapping> const& selected_stripes,
-    rmm::cuda_stream_view stream);
-
- private:
-  rmm::mr::device_memory_resource* _mr = nullptr;
-  std::vector<std::unique_ptr<datasource>> _sources;
+  std::vector<std::unique_ptr<datasource>> const _sources;  // Unused but owns data for others
   cudf::io::orc::detail::aggregate_orc_metadata _metadata;
-  cudf::io::orc::detail::column_hierarchy selected_columns;
+  cudf::io::orc::detail::column_hierarchy const _selected_columns;  // Need to be after _metadata
 
-  bool _use_index{true};
-  bool _use_np_dtypes{true};
-  std::vector<std::string> decimal128_columns;
-  data_type _timestamp_type{type_id::EMPTY};
-  reader_column_meta _col_meta{};
+  data_type const _timestamp_type;  // Override output timestamp resolution
+  bool const _use_index;            // Enable or disable attempt to use row index for parsing
+  bool const _use_np_dtypes;        // Enable or disable the conversion to numpy-compatible dtypes
+  std::vector<std::string> const _decimal128_columns;   // Control decimals conversion
```

Comment:
I'm mildly against const members as they impact copy/move. I'm not sure if there's a guideline on this.

Reply:
That will prevent copying, but not moving. I think we should not allow copying but we currently don't do that explicitly so leaving `const` members is just fine. In addition, members storing input parameters will not be changed thus marking them as `const` should be better to communicate their intention (store just unchanged input parameters).

Reply:
AFAIK it's the other way around. An object can't be moved from when it has const members because move generally modifies the members (e.g. moving a vector leaves the moved from object empty). So const members prevent moves, and objects have to be copied instead. Not a big reason, but still makes the situation iffy.